src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpResponseStream.cs

```diff
@@ -227,5 +243,33 @@ private void CheckDisposed()
                 throw new ObjectDisposedException(this.GetType().FullName);
             }
         }
+        
+        // The only way to abort pending async operations in WinHTTP is to close the request handle.
+        // This causes WinHTTP to cancel any pending I/O and accelerating its callbacks on the handle.
+        // This causes our related TaskCompletionSource objects to move to a terminal state.
+        //
+        // We only want to dispose the handle if we are actually waiting for a pending WinHTTP I/O to complete,
+        // meaning that we are await'ing for a Task to complete. While we could simply call dispose without
+        // a pending operation, it would cause random failures in the other threads when we expect a valid handle.
+        private void CancelPendingResponseStreamReadOperation()
```

Comment:
I think you're worried about this:

Is there a race condition between `WinHttpResponseStream.CancelPendingResponseStreamReadOperation` and `WinHttpRequstState.DisposeCtrReadFromResponseStream` ?

I believe that this two methods are exclusive already in terms of `CancelPendingResponseStreamReadOperation` actually calling _requestHandle.Dispose(). That is because it checks for (!_state.TcsQueryDataAvailable.Task.IsCompleted) before calling Dispose().  And if that check is true, it means that the operation isn't done. And DisposeCtrReadFromResponseStream() is only called when the operation is completed.

So, I think these are synchronized already by the thread-safe TaskCompletionSource<T>.